providers/os/provider/provider.go

```diff
@@ -319,6 +319,7 @@ func (s *Service) connect(req *plugin.ConnectReq, callback plugin.ProviderCallba
 		if len(idDetectors) == 0 {
 			// fallback to default id detectors
 			idDetectors = []string{ids.IdDetector_Hostname, ids.IdDetector_CloudDetect}
+			asset.IdDetector = idDetectors
```

Comment:
why do we need to set this here to the asset proto? can we build a test case for that behavior? if we need to set this, I would prefer to do this as part of the fingerprint setting.

Reply:
Testing this isn't that easy. We would need some sort of high level tests for this. The problem only happens when used with plugins. As soon as we use builtins, the problem disappears.

I'm not sure what you mean by "fingerprint setting". The detectors are what is used to determine the platform ID, so doing the fingerprinting. In my understanding they are the setting.
And we also need to set it inside the asset, because we later need them again. My current assumption is, this is needed because connect and discover are currently not handled separatly.

Reply:
This does not depend on MQL, therefore we can just test it within the provider. We just need to connect to an asset, run detection and disconnect and reconnect. See https://github.com/mondoohq/cnquery/blob/main/providers/vsphere/provider/provider_test.go#L19. 

We set multiple asset fields like name and platform ids here https://github.com/mondoohq/cnquery/pull/2563/files#diff-2ca4387e72bfea051603beafed7cf18d56fd372d0b3ada6287eb91de2ddf5b8aR385, why not set the id detectors there as well?

Reply:
Thanks for the pointers.

I refactored the fingerprinting and added a test for the idDetectors. But this test only tests the current code. It wouldn't have caught the issue this PR solves because the gRPC part is missing.